### ISSUE TITLE
Setting to toggle "error: Type of parameter $X must be a supertype of its class"

### ISSUE THREAD

**Issue by C-Loftus** [microsoft/pyright#7513]
**Is your feature request related to a problem? Please describe.**

There is no specific bug, however I believe a new setting would improve UX across different Python development setups.

**Describe the solution you’d like**

I believe that it would be helpful for users to separate `"error: Type of parameter $X must be a supertype of its class"` into it's own category or provide a setting to change it to a warning instead of an error. 
- Perhaps there could be a setting like `reportFirstParameterAsClassSubtype`  = true/false
- Without this we need to disable `(reportGeneralTypeIssues)` and lose many of Pyright's benefits

## Context 
I am a community member and developer in the [Talon Voice](https://talonvoice.com/) and  [Cursorless](https://github.com/cursorless-dev/cursorless) voice programming communities. Talon is configured with Python by using `action classes`, which are groups of functions within a class. The functions do not include `self` since the classes are not instantiated but are not static methods either since they are read in via the abstraction of the class decorator.

[Cursorless](https://github.com/cursorless-dev/cursorless) is a one large project that uses this pattern. 

## Example 

The following code is used in production for Cursorless and stylistically correct in our environment. However it throws errors with Pyright: 

`error: Type of parameter "key" must be a supertype of its class "Actions" (reportGeneralTypeIssues)` 
```
from talon import Module, actions, settings
mod = Module()

@mod.action_class
class Actions:

    def vscode_get_setting(key: str, default_value: Any = None):
        """Get the value of vscode setting at the given key"""
        path: Path = actions.user.vscode_settings_path()
        settings: dict = loads(path.read_text())

        if default_value is not None:
            return settings.get(key, default_value)
        else:
            return settings[key]
```

**Comment by erictraut** [comment]
Why do you not mark these as static methods? Is this required for correctness or just a choice to deviate from standard practices in Python (e.g. to save some keystrokes)?

**Comment by C-Loftus** [comment]
We don't use use static methods since it is more verbose for user configuration and doesn't semantically reflect what is going on. i.e. the functions are not necessarily associated with the class object at runtime but are just grouped with them for the purpose of a shared visual style when configuring.

**Comment by erictraut** [comment]
Hmm, this strikes me as an abuse of the language and the type system. You're declaring the method as an instance method, but then you're using behind-the-scenes magic to treat it otherwise. 

I see a few options here:
1. Do nothing. Pyright needs to make assumptions about how Python works, and if libraries stray too far away from the norms, static type checking will not work well. This feels to me like it might fall into this category. I don't know how big the cursorless community is, but it looks like its github page doesn't have that many stars (currently 1K), and [cursorless-talon](https://github.com/cursorless-dev/cursorless-talon) has fewer than 100 stars.

2. Do as you suggest and add a new diagnostic rule. We have done this occasionally in the past, but we require a good reason to do so because each new rule requires documentation, ongoing maintenance, and (when initially introduced) presents a breaking change for some pyright users. For this reason, we add new diagnostic rules reluctantly. This solution is also dissatisfying because it disables an otherwise-useful check unconditionally even if you want it to be suppressed only for specific classes.

3. Find some other means of statically differentiating between the normal case where an instance method should be treated as such and the case where it is not intended to be used as an instance method. Normally, a `@staticmethod` decorator is used for this. Is there another criterion we could use? One option is to base it on the name of the first parameter. If it's something other than `self` (or `cls` in a metaclass), then we suppress the type enforcement for the first parameter. It appears this might be what mypy is doing.

Right now, my inclination is to go with option 1 and reject this enhancement request unless/until we receive feedback from more pyright users that this is important to them.

However, I would value advice from other people in the Python typing community. @JelleZijlstra or @gvanrossum, do you guys have any thoughts about the above usage of instance methods? Have you seen this before? Would you expect a type checker to accommodate this usage?

**Comment by JelleZijlstra** [comment]
I agree this pattern is unusual and type checkers should not be expected to support it.

I could imagine ways to extend the type system to support the pattern; e.g., a decorator that makes it so the functions in a class body are not treated as methods and the class is not instantiable. However, given how rarely this appears to come up (I don't remember another use case in years of watching various typing-related trackers), I am not sure this issue is worth the cost of adding a new core type system feature.

As a workaround, could you add the `@staticmethod` decorator to your functions? That should do what you want, and also communicate the intent more clearly to users of the code.

**Comment by C-Loftus** [comment]
> should not be expected to support it.

Ok that is fair. I understand if it is out of scope. 

For context, in Talon scripting, it is an anti-pattern to decorate with `@staticmethod`.  That would imply it can be called by invoking the method on the class, but the user is _not intended_ to interact with the class in any way. To achieve contextual command overrides, there is a central namespace from which function implementations are dynamically determined at runtime based on context matching. 

Talon doesn't want the user to try and read from the class since it won't be the proper implementation or dynamically overridden correctly.

That is just the context if you are interested, I completely understand that it is certainly much different than most Python applications.

> It appears this might be what mypy is doing

Do you think mypy might be a better fit for situation?

**Comment by Azureblade3808** [comment]
Deriving from `Any` would suppress the typing issue --

```python
# pyright: reportSelfClsParameterName=false

from typing import Any

class Actions(Any):
    def vscode_get_setting(key: str, default_value: Any = None):  # No error
        ...
```

Is this a feasible work-around for you?

**Comment by C-Loftus** [comment]
> Is this a feasible work-around for you?

Unfortunately not, but that was a great idea.  In Talon, it throws a `TypeError: @action_class must not have a parent class`. Generally given the fact that the 'classes' in Talon are moreso groups of related functions and not acting like normal Python classes, it tries to discourage typing class-based design patterns like deriving from a parent class.

**Comment by Azureblade3808** [comment]
Then what about this --
```python
# pyright: reportSelfClsParameterName=false

from typing import Any, TYPE_CHECKING

if TYPE_CHECKING:
    Base = Any
else:
    Base = object

class Actions(Base):
    def vscode_get_setting(key: str, default_value: Any = None):
        ...
```
?

**Comment by erictraut** [comment]
Thanks for the discussion.

For now, I'm going to reject this feature request. I'm willing to revisit it in the future if there's additional signal from pyright users that this is a common pattern. 

I would encourage authors of libraries not to use this pattern because it deviates pretty significantly from the norms in Python and will inevitably lead to problems for static type checkers, linters, runtime type checkers, etc.

**Comment by C-Loftus** [comment]
> Then what about this --
Yes that does work, thanks! Didn't think of that.

> For now, I'm going to reject this feature request. I'm willing to revisit it in the future if there's additional signal from pyright users that this is a common pattern.
That is fair, thank you for discussion as well.
